Don't accept announcements for recently-closed channels.

rustyrussell opened the pull request:
Dragging in `gossipfrom` (now `gossipwith`) is a hack, but it's hard to test this without writing significant new infrastructure...

cdecker commented on tests/test_gossip.py:
```diff
@@ -876,13 +882,34 @@ def test_gossip_notices_close(node_factory, bitcoind):
+    # Grab channel_announcement from io logs (ends in ')
+    channel_announcement = l1.daemon.is_in_log('\[IN\] 0100').split(' ')[-1][:-1]
+    channel_update = l1.daemon.is_in_log('\[IN\] 0102').split(' ')[-1][:-1]
+    node_announcement = l1.daemon.is_in_log('\[IN\] 0101').split(' ')[-1][:-1]
```
Are those `[:-1]` to strip a newline? If yes, please use `.strip()` instead, that's more explicit.

cdecker commented on lightningd/gossip_control.c:
```diff
@@ -71,13 +71,13 @@ static void get_txout(struct subd *gossip, const u8 *msg)
 	op = wallet_outpoint_for_scid(gossip->ld->wallet, scid, scid);
 
-	if (op) {
+	if (op && !op->spendheight) {
```
We could also just push that down to the sqlite query by adding an argument `includespent`. That'd allow us to skip materializing the result in memory just to drop it again immediately.

cdecker replied:
Also `op` is allocated off of `scid` which could be a long-lived object (it currently isn't).

cdecker commented:
Just some minor comments, I'd like to push the spentness check into the query, since it makes the checks easier and avoids allocating memory if we don't need it.

rustyrussell commented:
Agreed.  Rewrite coming...

cdecker commented on wallet/wallet.c:
```diff
@@ -2139,7 +2139,8 @@ struct outpoint *wallet_outpoint_for_scid(struct wallet *w, tal_t *ctx,
 			  " AND txindex = ?"
-			  " AND outnum = ?");
+			  " AND outnum = ?"
+			  " AND spendheight IS NOT NULL");
```
Isn't this inverted? We're interested in the ones that aren't spent yet.

rustyrussell replied:
That would explain the test breakage :)

rustyrussell commented:
1. Fixed the inverted case.  Thanks!
2. Folded the two patches together, so there's no 'add code in caller, then remove it again to do it in db'.
3. Rebased on master to avoid spurious unrelated travis error.

cdecker commented:
Nice work @rustyrussell :rocket: 

ACK 64d70e182c4e8bdb719c42c69bb96b4dd69b02e0
